src/main/java/de/tum/in/www1/artemis/service/notifications/push_notifications/FirebasePushNotificationService.java

```diff
@@ -0,0 +1,99 @@
+package de.tum.in.www1.artemis.service.notifications.push_notifications;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executors;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.retry.RetryCallback;
+import org.springframework.retry.support.RetryTemplate;
+import org.springframework.scheduling.annotation.Async;
+import org.springframework.scheduling.annotation.EnableAsync;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+import com.google.common.collect.Lists;
+import com.google.gson.Gson;
+
+import de.tum.in.www1.artemis.domain.push_notification.PushNotificationDeviceType;
+import de.tum.in.www1.artemis.repository.PushNotificationDeviceConfigurationRepository;
+
+/**
+ * Handles the sending of Android Notifications to the Relay Service
+ */
+@Service
+@EnableAsync(proxyTargetClass = true)
+public class FirebasePushNotificationService extends PushNotificationService {
+
+    private final Logger log = LoggerFactory.getLogger(FirebasePushNotificationService.class);
+
+    private final PushNotificationDeviceConfigurationRepository repository;
+
+    @Value("${artemis.push-notification-relay:#{Optional.empty()}}")
+    private Optional<String> relayServerBaseUrl;
+
+    private final RestTemplate restTemplate;
+
+    public FirebasePushNotificationService(PushNotificationDeviceConfigurationRepository pushNotificationDeviceConfigurationRepository, RestTemplate restTemplate) {
+        repository = pushNotificationDeviceConfigurationRepository;
+        this.restTemplate = restTemplate;
+    }
+
+    @Override
+    void sendNotificationRequestsToEndpoint(List<RelayNotificationRequest> requests, String relayBaseUrl) {
```

Comment:
A lot of this method and the next one is pretty much identical to the `ApplePushNotificationService`. Can you refactor it out into a `protected` method of the shared base class?